File: ImageCompression/ImgCompression.py

"""
The next code aims to perform image compression using the proposed
Singular Value Decomposition (SVD) method
"""

import os

import matplotlib.pyplot as plt
import numpy as np
from PIL import Image


# Upload an image
def UploadImg(img_name):
    script_directory = os.path.dirname(os.path.abspath(__file__))
    image_filename = img_name
    image_path = os.path.join(script_directory, image_filename)
    img = Image.open(image_path)
    return img

# Get image dimensions, width, and height
def GetImageSize(img):
    width, height = img.size
    print("Image width:", width)
    print("Image height:", height)

# Normalize image from 0-255 to 0 to 1
def NormalizedImg(img):
    img_array = np.array(img)
    normalized_img_array = img_array / 255.0
    return normalized_img_array

# Store Image in the current directory
def StoreImg(normalized_img_array, file_name):
    script_directory = os.path.dirname(os.path.abspath(__file__))
    normalized_image_path = os.path.join(script_directory, file_name)
    normalized_img_pil = Image.fromarray((normalized_img_array * 255).astype(np.uint8))
    normalized_img_pil.save(normalized_image_path)

# Function to write content to a file in the same folder as the script
def write_to_file(file_name, content):
    try:
        # Get the directory path of the current script
        script_dir = os.path.dirname(os.path.abspath(__file__))
        
        # Combine the script directory with the provided file name
        file_path = os.path.join(script_dir, file_name)

        # Open the file in write mode
        with open(file_path, 'w') as file:
            # If the file exists and is not empty, clear its content
            if os.path.exists(file_path) and os.path.getsize(file_path) > 0:
                file.truncate(0)
            
            # Write the new content to the file
            file.write(content)

        #print(f"Content written to {file_path}")
    except Exception as e:
        print(f"An error occurred: {str(e)}")

def get_file_length(file_name):
    try:
        # Get the directory path of the current script
        script_dir = os.path.dirname(os.path.abspath(__file__))
        
        # Combine the script directory with the provided file name
        file_path = os.path.join(script_dir, file_name)

        # Read the content from the file and split it into rows
        with open(file_path, 'r') as file:
            lines = file.readlines()

        # Calculate matrix dimensions (rows x columns)
        rows = len(lines)
        if rows > 0:
            cols = len(lines[0].strip().split(','))
            print(f"Matrix dimensions: {rows} rows x {cols} columns")
        else:
            print("Matrix is empty.")

    except FileNotFoundError:
        print(f"File not found: {file_name}")
    except Exception as e:
        print(f"An error occurred: {e}")



# Load matrices A and B
def LoadMatrices(file_name):
    try:
        # Get the directory path of the current script
        script_dir = os.path.dirname(os.path.abspath(__file__))
        
        # Combine the script directory with the provided file name
        file_path = os.path.join(script_dir, file_name)

        # Read matrices from the file
        matrices = np.loadtxt(file_path)
        return matrices
    except Exception as e:
        #print(f"An error occurred: {str(e)}")
        return None

# The next section contains the implementation of the SVD algorithm 
def svdPython(A):

    m, n = A.shape         

    if m >= n:              
        M1 = A.T @ A        
        y_1, v_1 = np.linalg.eigh(M1)  
        const = n * np.max(y_1) * np.finfo(float).eps 
        y_2 = (y_1 > const) 
        r_A = np.sum(y_2)   
        y_3 = np.abs(y_1) * y_2 
        s_1, orden = np.sort(np.sqrt(y_3), axis=0)[::-1], np.argsort(np.sqrt(y_3), axis=0)[::-1]
        v_2 = v_1[:, orden] 
        v = v_2[:, :r_A]    
        s = np.diag(s_1[:r_A])  
        u = (1. / s_1[:r_A]) * (A @ v) 

    else:
        M1 = A @ A.T 
        y_1, u_1 = np.linalg.eigh(M1)   
        const = m * np.max(y_1) * np.finfo(float).eps  
        y_2 = (y_1 > const)  
        r_A = np.sum(y_2)    
        y_3 = np.abs(y_1) * y_2 
        s_1, orden = np.sort(np.sqrt(y_3), axis=0)[::-1], np.argsort(np.sqrt(y_3), axis=0)[::-1]
        u_2 = u_1[:, orden]  
        u = u_2[:, :r_A]     
        s = np.diag(s_1[:r_A])  
        v = (1. / s_1[:r_A]) * (A.T @ u) 
        
    return u, s, v

def ImageCompression():

    # Upload the original image
    img = UploadImg('img1.jpg')
    
    # Normalize the image
    normalized_img = NormalizedImg(img)
    
    # Display the original image
    plt.subplot(1, 3, 1)
    plt.imshow(normalized_img)
    plt.title('Original Image')
    
    # Store the normalized image
    StoreImg(normalized_img, 'normalized_image.jpg')

    # Write the normalized image to a text file
    normalized_image_txt = '\n'.join([','.join(map(str, row)) for row in normalized_img])
    write_to_file('originalImage.txt', normalized_image_txt)

    print(get_file_length('originalImage.txt'))

    # Perform image compression using SVD
    r = 10
    u, s, v = svdPython(normalized_img)
    Ur = u[:, :r]
    Sr = s[:r, :r]
    Vr = v[:, :r]
    D = Ur @ Sr
    C = Vr.T
    Ic = D @ C

    # Display the compressed image
    plt.subplot(1, 3, 2)
    plt.imshow(Ic)
    plt.title('Compressed Image')
    
    # Store matrices D and C
    write_to_file('D.txt', '\n'.join([','.join(map(str, row)) for row in D]))
    write_to_file('C.txt', '\n'.join([','.join(map(str, row)) for row in C]))

    print("D")
    print(get_file_length('D.txt'))
    print("C")
    print(get_file_length('C.txt'))
    
    # Load matrices D and C
    D1 = LoadMatrices("D.txt")
    C1 = LoadMatrices("C.txt")

    if D1 is not None and C1 is not None:
        Ic1 = D1 @ C1

        # Display the compressed image using loaded matrices D and C
        plt.subplot(1, 3, 3)
        plt.imshow(Ic1)
        plt.title('Compressed Image (Loading matrices D and C)')

    plt.show()
    
ImageCompression()

File: Julia/ImageCompression/Img_Compression.jl
using FilePaths
using Images
using LinearAlgebra
using BenchmarkTools
using DelimitedFiles 
using FileIO
using ColorTypes

# Function to open an image in the same directory as the script
function open_image(filename)
    script_path = abspath(dirname(@__FILE__))
    image_path = joinpath(script_path, filename)
    return load(image_path)
end

# Function to get the width and height of an image
function get_image_dimensions(image)
    return size(image)
end

function normalize_image(img::AbstractMatrix{<:Unsigned})
    # Convert the input image to Float64 and scale to [0, 1]
    normalized_img_array = Float64.(img) / 255.0

    return normalized_img_array
end

function scale_gray_to_0_255(image::AbstractMatrix{<:Gray{N0f8}})
    # Convert the image to UInt8 directly
    scaled_image = reinterpret(UInt8, image)
    return scaled_image
end

# Function to store an image as a PNG file
function store_image_as_png(image, filename)
    script_path = abspath(dirname(@__FILE__))
    output_path = joinpath(script_path, filename)
    
    try
        # Save the image as a PNG file
        save(output_path, image, format="jpeg")
        println("Image saved to $output_path")
    catch e
        println("An error occurred while saving the image: $e")
    end
end

# Function to write a matrix to a file in the same folder as the script
function write_matrix_to_file(file_name::AbstractString, matrix::AbstractMatrix{T}) where T
    try
        # Get the directory path of the current script
        script_dir = dirname(@__FILE__)

        # Combine the script directory with the provided file name
        file_path = joinpath(script_dir, file_name)
        
        # Check if the directory exists, create it if it doesn't
        if !isdir(script_dir)
            mkdir(script_dir)
        end

        # Open the file in write mode
        open(file_path, "w") do file
            # If the file exists and is not empty, clear its content
            if isfile(file_path) && filesize(file_path) > 0
                truncate(file, 0)
            end

            # Write the new content to the file
            for row in eachrow(matrix)
                println(file, join(row, ", "))
            end
        end

        println("Content written to $file_path")
    catch e
        println("An error occurred: $e")
    end
end

# Function to read a matrix from a file in the same folder as the script
function read_matrix_from_file(file_name::AbstractString)
    try
        # Get the directory path of the current script
        script_dir = dirname(@__FILE__)

        # Combine the script directory with the provided file name
        file_path = joinpath(script_dir, file_name)

        # Read the matrix from the file using readdlm from DelimitedFiles
        return DelimitedFiles.readdlm(file_path, ',', Float64)
    catch e
        # println("An error occurred: $e")
        return nothing
    end
end

function get_file_length(file_name::AbstractString)
    try
        # Get the directory path of the current script
        script_dir = dirname(@__FILE__)
        
        # Combine the script directory with the provided file name
        file_path = joinpath(script_dir, file_name)

        # Read the content from the file and split it into rows
        lines = readlines(file_path)

        # Calculate matrix dimensions (rows x columns)
        rows = length(lines)
        if rows > 0
            cols = length(split(strip(lines[1]), ','))
            println("Matrix dimensions: $rows rows x $cols columns")
        else
            println("Matrix is empty.")
        end

    catch e
        println("An error occurred: $e")
    end
end

function svdJulia(A)
    n = size(A, 2)
    m = size(A, 1)

    if m >= n
        M1 = A' * A
        y_1, v_1 = eigen(M1)
        consta = n * maximum(y_1) * eps(Float64)
        y_2 = y_1 .> consta
        r_A = sum(y_2)
        y_3 = abs.(y_1) .* y_2
        s_1, orden = sort(sqrt.(y_3), rev=true), sortperm(sqrt.(y_3), rev=true)
        v_2 = v_1[:, orden]
        v = v_2[:, 1:r_A]
        s = Diagonal(s_1[1:r_A])
        u = A * v / s
    else
        M1 = A * A'
        y_1, u_1 = eigen(M1)
        consta = m * maximum(y_1) * eps(Float64)
        y_2 = y_1 .> consta
        r_A = sum(y_2)
        y_3 = abs.(y_1) .* y_2
        s_1, orden = sort(sqrt.(y_3), rev=true), sortperm(sqrt.(y_3), rev=true)
        u_2 = u_1[:, orden]
        u = u_2[:, 1:r_A]
        s = Diagonal(s_1[1:r_A])
        v = A' * u / s
    end

    return u, s, v
end


# Function to generate an image from a matrix in a text file
function generate_image_from_txt(txt_file::AbstractString, output_image::AbstractString)
    # Get the directory path of the current script
    script_path = abspath(dirname(@__FILE__))

    # Combine the script directory with the provided input and output file names
    input_txt_path = joinpath(script_path, txt_file)
    output_image_path = joinpath(script_path, output_image)

    # Read the matrix from the text file in the same directory
    matrix = readdlm(input_txt_path, ',', Float64)

    # Convert the matrix to an image
    img = Gray.(matrix)  # Assuming a grayscale image, change to RGB if needed

    # Save the image as a JPEG file in the same directory as the script
    save(output_image_path, img)
    
    println("Image saved to $output_image_path")
end

function transform_to_0_255_scale(input_filename::AbstractString, output_filename::AbstractString)
    # Get the directory path of the current script
    script_dir = dirname(@__FILE__)
    
    # Combine the script directory with the provided file names
    input_path = joinpath(script_dir, input_filename)
    output_path = joinpath(script_dir, output_filename)
    
    try
        # Read the matrix from the input text file
        matrix = readdlm(input_path, ',', Float64)

        # Scale the values from [0, 1] to [0, 255]
        scaled_matrix = round.(Int, matrix .* 255)

        # Write the scaled matrix to the output text file
        writedlm(output_path, scaled_matrix, ',')

        println("Scaled matrix saved to $output_path")
    catch e
        println("An error occurred: $e")
    end
end


# Function to read a text file with integer values in [0-255] scale and transform them into an image
function read_text_file_to_image(txt_file::AbstractString, output_image::AbstractString)
    # Get the directory path of the current script
    script_dir = dirname(@__FILE__)

    # Combine the script directory with the provided file names
    txt_file_path = joinpath(script_dir, txt_file)
    output_image_path = joinpath(script_dir, "$output_image.png")  # Specify PNG format here

    # Read the matrix from the text file as integers
    matrix = readdlm(txt_file_path, ',', Int)

    # Convert the matrix to Gray{N0f8} type with values in [0, 1] scale
    matrix = Gray.(matrix ./ 255.0)

    # Create an image from the matrix
    img = Gray.(matrix)

    # Save the image in the same directory as the script with the specified filename
    save(output_image_path, img)

    println("Image saved to $output_image_path")
end

# Main code
function ImageCompression()

    # Open the image
    I1 = open_image("img1.jpg")

    I2 = scale_gray_to_0_255(I1)

    # Normalize and convert to Matrix{RGB}
    I = normalize_image(I2)

    # Extract the pixel data and convert to a suitable format for writing to a text file
    pixel_data = channelview(I)

    # Save the pixel data to a text file
    write_matrix_to_file("originalImage.txt", pixel_data)

    # Create compressed data
    r = 100
    U, S, V = svdJulia(pixel_data)
    Ur = U[:, 1:r]
    Sr = S[1:r, 1:r]
    Vr = V[:, 1:r]
    D = Ur * Sr
    C = Vr'
    Ic = D * C

    # Save the normalized image as a PNG file
    # store_image_as_png(Ic, "compressed_image.jpg")

    # Save matrices D and C to files
    write_matrix_to_file("C.txt", C)
    write_matrix_to_file("D.txt", D)

    # Load matrices D and C
    D1 = read_matrix_from_file("D.txt")
    C1 = read_matrix_from_file("C.txt")
    
    if D1 === nothing || C1 === nothing
        println("Error: Failed to load matrices D and C.")
        return
    end

    Ic1 = (D1 * C1) 

    write_matrix_to_file("Ic1.txt", Ic1)
   
    transform_to_0_255_scale("Ic1.txt", "O2.txt")

    read_text_file_to_image("O2.txt", "compressed_image_100")
end

ImageCompression()

